### PR TITLE
post: h3 hexes

### PR DESCRIPTION
**Pull request by jacobdadams**
Ok, this has a much cleaner git history now and is ready for review (I think). Feedback and critiques are welcome.

### CODE REVIEW

File: _posts/2022-10-25-using-h3-hexes.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+{: .flex .flex--center}
+
+The important point is that an individual hex ID contains the IDs of the hexes "above" it in the hierarchy, kind of like how census GeoIDs contain the block, tract, county, and state codes. This makes it easy to traverse up to a larger hex without re-calculating the IDs.
+
+Using these multiple scales (which H3 calls resolutions), we can **create hexes at multiple scales to aggregate our data to larger and larger areas.** With these, we can view trends at the neighborhood, city, or county level. We can also create web maps that allow us to drill down from general trends to neighborhood or block level to see what areas contribute the most.
+
+### Assigning H3 IDs is Fast
+{: .text-left}
+
+As noted above, H3 uses some fancy mathematics under the hood to assign hex ids instead of doing spatial analysis. Written in C and exposed through a variety of bindings in other languages, these calculations are **pretty darn fast at figuring out what hex cell a point belongs in.**
```
Comment (steveoh) [inline _posts/2022-10-25-using-h3-hexes.md:59]: **question**: is this math done on the client side or does it use an api that could disappear?

**Review by steveoh — APPROVED** [review]
Bravo. Love it.

File: _posts/2022-10-25-using-h3-hexes.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+For example, I used `%timeit` to  assign a resolution 9 ID to all 1,339,635 of our [address points]({% link data/location/address-data/index.html %}) (after projecting them to WGS84 to get direct access to lat/long in the SHAPE field):
+
+```python
+def assign_h3(df, resolution):
+    df[f'h3_{resolution}'] = df.apply(lambda row: h3.h3_to_string(h3.geo_to_h3(row['SHAPE']['y'], row['SHAPE']['x'], resolution)), axis=1)
+
+addr_df = pd.DataFrame.spatial.from_featureclass(r'C:\gis\Projects\H3\H3.gdb\address_points_20220727_wgs84')
```
Comment (steveoh) [inline _posts/2022-10-25-using-h3-hexes.md:67]: **suggestion**: use a relative path to remove all the c:\ noise on all the paths used

File: _posts/2022-10-25-using-h3-hexes.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+#: Row index 0 is a mask outside the state boundary, index 1 is the boundary itself.
+#: Also, the SHAPE field is the last item in the column index
+buffered = state_boundary_df.iloc[1, -1].buffer(5000)
+buffered_df = pd.DataFrame.spatial.from_df(
+                pd.DataFrame({'SHAPE': [buffered]}), 
+                geometry_column='SHAPE', sr=26912
+                )
+
+#: Project our polygon to WGS84 and convert to geojson for the H3 analysis
+buffered_df.spatial.project(4326)
```
Comment (steveoh) [inline _posts/2022-10-25-using-h3-hexes.md:105]: **question**: does the project command allow you to use a transformation?

**Review by steveoh — DISMISSED** [review]
just a few more comments

**Review by stdavis — APPROVED** [review]
This is awesome and I couldn't find anything to give feedback on.

File: _posts/2022-10-25-using-h3-hexes.md (hunk 1 above)
Comment (jacobdadams) [reply _posts/2022-10-25-using-h3-hexes.md:59]: I'm 95% sure it's all local. I haven't found an explicit statement that is the case, but there's a ton of C code in the source. My C is rusty (no pun intended) but I'm pretty sure the algorithm implementation is in there.

File: _posts/2022-10-25-using-h3-hexes.md (hunk 3 above)
Comment (jacobdadams) [reply _posts/2022-10-25-using-h3-hexes.md:105]: It does, but there's apparently a bug in the way the ArcGIS API for Python calls the underlying functions that causes a serious time disadvantage, making it impossible to project anything more than a very small spatial dataframe. So I'm just ignoring it here.

**Review by steveoh — APPROVED** [review]

File: _posts/2022-10-25-using-h3-hexes.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
+---
+title: 'Aggregating and Analyzing Point Data with H3 Hexes and Pandas'
+author:
+  display_name: Jake Adams
+  email: jdadams@utah.gov
+date: 2022-10-25 12:20:21
```
Comment (steveoh) [inline _posts/2022-10-25-using-h3-hexes.md:6]: **chore**: update this for tomorrow and we can publish it

**Comment by nathankota** [comment]
Two thumbs up! Well done!

**Review by eneemann — COMMENTED** [review]
I like it!  I think it's ready to go.

**Review by steveoh — APPROVED** [review]

**Pull request merged**